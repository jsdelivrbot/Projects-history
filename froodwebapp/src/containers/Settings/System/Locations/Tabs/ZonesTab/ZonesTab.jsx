import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Row, Col } from 'react-flexbox-grid';
import {
  locationZonesGetRequest,
  locationZonesSaveRequest,
  locationZonesUpdateRequest,
} from 'redux-base/actions';
import { Table, Spin, Select } from 'antd';
import { ConnectedTopFormModal, ActionButton } from 'components';
import { table, select } from 'styles/common.scss';
import columns from './zonesTabHelpers';
import fields from './modalFields';

const mapStateToProps = state => ({
  locations: state.locations.locations,
  locationZones: state.locations.locationZones,
  loadingPage: state.locations.loadingPage,
  needReloadLocationZones: state.locations.needReloadLocationZones,
  zoneTypes: state.commonData.zoneTypes,
});

const mapDispatchToProps = {
  locationZonesGetRequest,
  locationZonesSaveRequest,
  locationZonesUpdateRequest,
};

const newZone = {
  name: '',
  type: ''
};

export class ZonesTab extends Component {
  constructor(props) {
    super(props);
    this.state = {
      modalVisible: false,
      modalData: newZone,
      selectedLocationId: ''
    };
  }

  componentWillReceiveProps(nextProps) {
    // when we update zone
    if (nextProps.needReloadLocationZones) {
      this.setState({
        modalVisible: false,
        modalData: undefined
      }, () => this.props.locationZonesGetRequest({
        id: this.state.selectedLocationId
      }));
      return;
    }

    // when we create zone
    if (this.props.locationZones.length !== nextProps.locationZones.length) {
      this.setState({
        modalVisible: false,
        modalData: newZone
      });
    }
  }

  handleToggleModal = () => {
    this.setState({
      modalVisible: !this.state.modalVisible,
      modalData: newZone
    });
  }

  handleSave = (payload) => {
    if (this.state.modalData.id) {
      this.props.locationZonesUpdateRequest({
        id: this.state.selectedLocationId,
        payload: {
          id: payload.id,
          name: payload.name
        }
      });
    } else {
      this.props.locationZonesSaveRequest({
        id: this.state.selectedLocationId,
        payload
      });
    }
  }

  handleLocationSelect = (id) => {
    this.setState({
      selectedLocationId: id
    }, () => {
      this.props.locationZonesGetRequest({
        id
      });
    });
  }

  handleEdit = (e) => {
    this.setState({
      modalVisible: true,
      modalData: this.props.locationZones.find(item => item.id === Number(e.target.id))
    });
  }

  render() {
    const {
      modalData,
      modalVisible,
      selectedLocationId
    } = this.state;

    const {
      locations,
      zoneTypes,
      locationZones,
      loadingPage
    } = this.props;
    return (
      <div>
        <ConnectedTopFormModal
          loading={ loadingPage }
          title="New Zone"
          buttonVisible={ false }
          handleSave={ this.handleSave }
          handleToggleModal={ this.handleToggleModal }
          visible={ modalVisible }
          initialValues={ modalData }
          fields={ fields(zoneTypes) }
        />
        <Row style={ { marginBottom: '1rem' } } middle="xs" between="xs">
          <Col xs md={ 6 } lg={ 4 }>
            <Select
              className={ select }
              placeholder="Location"
              onChange={ this.handleLocationSelect }
            >
              { locations.map(item => (
                <Select.Option
                  key={ item.id }
                  value={ item.id.toString() }
                >
                  { item.name }
                </Select.Option>
              ))}
            </Select>
          </Col>
          <Col lg>
            <ActionButton
              style={ { float: 'right' } }
              disabled={ !selectedLocationId }
              onClick={ this.handleToggleModal }
            >
              New Zone
            </ActionButton>
          </Col>
        </Row>
        <Spin spinning={ loadingPage }>
          <Table
            className={ table }
            rowKey="id"
            size="small"
            columns={ columns(this.handleEdit, this.handleDeactivate) }
            dataSource={ locationZones }
          />
        </Spin>
      </div>
    );
  }
}

ZonesTab.propTypes = {
  // trigger
  loadingPage: PropTypes.bool.isRequired,
  needReloadLocationZones: PropTypes.bool.isRequired,
  // data
  locations: PropTypes.array.isRequired,
  locationZones: PropTypes.array.isRequired,
  zoneTypes: PropTypes.array.isRequired,
  // redux-base
  locationZonesGetRequest: PropTypes.func.isRequired,
  locationZonesSaveRequest: PropTypes.func.isRequired,
  locationZonesUpdateRequest: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(ZonesTab);
