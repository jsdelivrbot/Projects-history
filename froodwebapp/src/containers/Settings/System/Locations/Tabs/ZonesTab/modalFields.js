export default zoneTypes => [{
  header: 'Zone Name',
  type: 'text',
  name: 'name',
  editableForExistingItem: true,
  editableForNewItem: true
}, {
  header: 'Zone Type',
  type: 'select',
  name: 'type',
  editableForExistingItem: false,
  editableForNewItem: true,
  menuItems: zoneTypes.map(item => ({ key: item.id, value: item.name }))
}];
