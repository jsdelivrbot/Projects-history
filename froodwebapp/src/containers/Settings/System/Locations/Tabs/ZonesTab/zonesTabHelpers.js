import React from 'react';
import { TableActionButtons } from 'components';

export default handleEdit => [{
  title: 'ZONE NAME',
  dataIndex: 'name'
}, {
  title: 'ZONE TYPE',
  dataIndex: 'typeDescription'
}, {
  title: 'ACTION',
  render: (text, record) => (
    <TableActionButtons
      record={ record }
      handleEdit={ handleEdit }
    />
  )
}];
