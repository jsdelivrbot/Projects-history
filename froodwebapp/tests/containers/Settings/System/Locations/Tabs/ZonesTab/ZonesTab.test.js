import { ZonesTab } from 'containers/Settings/System/Locations/Tabs/ZonesTab/ZonesTab';
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const { describe, expect, it } = global;

describe('ZonesTab', () => {
  const props = {
    loadingPage: false,
    needReloadLocationZones: false,
    locations: [],
    locationZones: [],
    zoneTypes: [],
    locationZonesGetRequest: jest.fn(),
    locationZonesSaveRequest: jest.fn(),
    locationZonesUpdateRequest: jest.fn()
  };

  it('renders the ZonesTab component', () => {
    const component = shallow(<ZonesTab { ...props } />);

    expect(component.length).toEqual(1);
  });

  it('capturing Snapshot of ZonesTab', () => {
    const component = shallow(<ZonesTab { ...props } />);
    const view = toJson(component);
    const actual = expect(view);

    actual.toMatchSnapshot();
  });

  it('calls locationZonesSaveRequest in handleSave when activeTab is locations', () => {
    const payload = {
      id: 'id',
      isActive: false,
      isDefault: false,
      holdStock: jest.fn()
    };
    const component = shallow(<ZonesTab { ...props } />);
    component.setState({ selectedLocationId: '1' });
    component.instance().handleSave(payload);
    expect(props.locationZonesSaveRequest.mock.calls.length).toBe(1);

    expect(props.locationZonesSaveRequest.mock.calls[0][0]).toEqual({
      id: '1',
      payload
    });
  });

  it('calls locationZonesGetRequest in handleLocationSelect when activeTab is locations', () => {
    const id = '1';
    const component = shallow(<ZonesTab { ...props } />);
    component.instance().handleLocationSelect(id);
    expect(props.locationZonesGetRequest.mock.calls.length).toBe(1);

    expect(props.locationZonesGetRequest.mock.calls[0][0]).toEqual({
      id
    });
  });
});
