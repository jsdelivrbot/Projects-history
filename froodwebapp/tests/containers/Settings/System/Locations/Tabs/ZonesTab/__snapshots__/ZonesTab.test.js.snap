// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZonesTab capturing Snapshot of ZonesTab 1`] = `
<div>
  <ReduxForm
    buttonVisible={false}
    fields={
      Array [
        Object {
          "editableForExistingItem": true,
          "header": "Zone Name",
          "name": "name",
          "type": "text",
        },
        Object {
          "editableForExistingItem": false,
          "header": "Zone Type",
          "menuItems": Array [],
          "name": "type",
          "type": "select",
        },
      ]
    }
    handleSave={[Function]}
    handleToggleModal={[Function]}
    initialValues={
      Object {
        "name": "",
        "type": "",
      }
    }
    loading={false}
    title="New Zone"
    visible={false}
  />
  <Row
    between="xs"
    middle="xs"
    style={
      Object {
        "marginBottom": "1rem",
      }
    }
  >
    <Col
      lg={4}
      md={6}
      xs={true}
    >
      <Select
        choiceTransitionName="zoom"
        className="select"
        onChange={[Function]}
        placeholder="Location"
        prefixCls="ant-select"
        showSearch={false}
        transitionName="slide-up"
      />
    </Col>
    <Col
      lg={true}
    >
      <ActionButton
        disabled={true}
        onClick={[Function]}
        style={
          Object {
            "float": "right",
          }
        }
      >
        New Zone
      </ActionButton>
    </Col>
  </Row>
  <Spin
    prefixCls="ant-spin"
    size="default"
    spinning={false}
    wrapperClassName=""
  >
    <Table
      bordered={false}
      className="table"
      columns={
        Array [
          Object {
            "dataIndex": "name",
            "title": "ZONE NAME",
          },
          Object {
            "dataIndex": "typeDescription",
            "title": "ZONE TYPE",
          },
          Object {
            "render": [Function],
            "title": "ACTION",
          },
        ]
      }
      dataSource={Array []}
      indentSize={20}
      loading={false}
      locale={Object {}}
      prefixCls="ant-table"
      rowKey="id"
      rowSelection={null}
      showHeader={true}
      size="small"
      useFixedHeader={false}
    />
  </Spin>
</div>
`;
